interface/ceed-elemrestriction.c

```diff
@@ -308,6 +308,54 @@ int CeedElemRestrictionApply(CeedElemRestriction rstr, CeedTransposeMode tmode,
   return 0;
 }
 
+/**
+  @brief Restrict an L-vector to a block of an E-vector or apply transpose
+
+  @param rstr    CeedElemRestriction
+  @param block   Block number to restrict to/from
+  @param tmode   Apply restriction or transpose
+  @param lmode   Ordering of the ncomp components
+  @param u       Input vector (of size @a ndof when tmode=CEED_NOTRANSPOSE)
+  @param v       Output vector (of size @a nelem * @a elemsize when tmode=CEED_NOTRANSPOSE)
+  @param request Request or CEED_REQUEST_IMMEDIATE
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Advanced
+**/
+int CeedElemRestrictionApplyBlock(CeedElemRestriction rstr, CeedInt block,
+                                  CeedTransposeMode tmode,
+                                  CeedTransposeMode lmode,
+                                  CeedVector u, CeedVector v,
+                                  CeedRequest *request) {
```

Comment:
Should the interface be designed to support a zero-copy implementation when element restriction is the identity?

Reply:
Two questions. 
A) Would this only be for single element processing? Multi element processing needs a copy to do the shuffle.
B) Would it be better to modify this interface or have the backend handle the special case of single element blocks with an identity restriction.

Reply:
I think this ordering is most relevant at high order where single-element processing is competitive. One could also imagine an E-vector with block/interlaced elements and an "identity" restriction in that ordering.

I'm concerned about public interfaces that may have poor granularity for some architectures. It lures users into writing fragile code that needs multiple branches for different hardware or for different application choices.

One option for zero-copy that does not change granularity is for the function to return the location where the E-vector data lives, thus having the option to return a pointer to the data where it lies, versus in a buffer.

Reply:
I agree that this is important for performance, but I'm trying to understand the best place to put the extra complexity.

Are you worried about a user calling `CeedElemRestrictionApply` in their code or a backend calling `CeedElemRestrictionApply`?

I don't see a way that an identity restriction in transpose mode avoids branching logic in the backend's `CeedOperatorApply` code. That decision reaches all the way back to the decision of where to write the output of `CeedBasisApply` after the QFunction, or possibly where to write the output of `CeedQFunctionApply` for `CEED_EVAL_NONE`. In notransopose mode we can, and at one point used to, make the evec point to the lvec data for inputs with an identity restriction (that was dropped in the current blocked backend because of the interlacing), so that side isn't as tricky.

I worry about an interface that returns a data location rather than a vector causing issues for the multi-memory model in different backends.

Currently we copy full evecs, even for identity restrictions. Does better handling of identity restrictions best fit in this PR or a follow-on?

Reply:
I'm mainly concerned about user code. We maintain all the backends in existence at this point, though we want to eventually have a stable way to support backends maintained externally.

With the current interface, libCEED could determine that the restriction is the identity and make the output vector alias the block of the input vector (instead of copying into separate memory). It could mark the result read-only, but would need some way of releasing the aliased vector. Perhaps a Get/Restore interface. For the transpose mode, we could Get a block of the E-vector for writing and Restore it when done.
* Identity restriction and matching memory spaces
  * Get sets a writeable view; Restore drops the view
* Otherwise
  * Get ensures a writeable buffer; Restore applies the transpose restriction.

Reply:
This is starting to sound like a follow-on PR in terms of scope. This PR already provides performance enchantment, so maybe we merge this one and start a new PR better handling identity restrictions?

Reply:
Yes, that's fine with me.